B-教学案例与实践/B13-AI对联生成案例/README.md

```diff
@@ -0,0 +1,492 @@
+
+在本次案例中，我们使用深度学习的方法，实现一个对联自动生成微信小程序 —— ***联景联情***
+
+可以在微信中搜索 “联景联情” 找到并使用该小程序，或者扫描如下二维码获取。
+
+![联景联情](./src/imgs/qrcode.jpg "扫码使用小程序")
+
+# 案例概要
+
+## 案例描述
+
+用户在小程序中轻松上传一张图像，程序提取图像信息，自动生成3组5-7字的备选上下联。用户可选择一组心仪的对联，与图像合成你的专属对联。
+
+该程序是由学习微软亚洲研究院首席研发经理邹欣老师《软件工程实践》课程的几位同学（他们均在微软亚洲研究院实习）组队完成的，在完成过程中充分利用课程所学软件工程知识，结合NLP知识和软件开发技能，完成了一个端到端的应用服务。这里可以看到他们的开发感想与总结。
+
+## 程序结构
+
+本案例的基本程序结构如下图所示：
+
+![程序结构图](./images/codeflow.PNG)
+
+后续将会对每个部分进行详细说明。
+
+## 涉及知识
+
+* 使用微软认知服务（Cognitive Service）中计算机视觉（computer vision）服务
+* NLP 相关知识
+   * Sequence2Sequence 模型
+   * 相关算法： RNN，LSTM，GRU，Transformer
+* 模型库的使用
+   * Tensor2Tensor
+   * Fairseq
+
+## 案例价值
+
+此案例特色显明，生动有趣，可以激发学生们对深度学习的兴趣。在技术层面，此案例使学生对深度学习的时序模型有直观的了解。该案例面向对象广泛，扩展性强。对初学者，可重复案例的过程；对于进阶者，不论在模型选择上，还是在模型推理上，都可以有更多的扩展，可提高学生们的探索研究能力。
+
+
+# 案例详解
+
+## 搭建环境
+
+### 证书许可
+
+由于微信小程序使用https协议访问服务器，需要申请SSL证书。本案例申请了微软的SSL证书。
+
+### 操作系统
+
+本案例运行在Azure虚拟机上，虚拟机的系统为Ubuntu 16.04
+
+### 编程语言
+
+本案例的语言使用 Python3.x ，并需要安装一些 python packages。
+
+1. 训练所需 python packages 在文件 [train_requirements.txt](./src/training/pip_requirerments.txt) 中。
+2. 服务所需 python packages 在文件 [conda_requirements.txt](./src/service/conda_requirements.txt) 中。
+
+### 框架和模型库
+
+本案例使用 tensorflow 的 tensor2tensor 模型库，具体版本如下：
+- tensorflow 1.4.0
+- tensor2tensor 1.2.9
+
+## 模型训练
+
+### 模型选择
+
+想要完成一个自动生成对联的小程序，想法十分美好，但想要达到这个目标，光拍拍脑袋想想是不够的，需要训练出一个能完成对联生成的自然语言理解模型。于是乎，就有两个选择：
+
+1. 自己写一套完成对联生成工作的深度学习模型。这个工作量相当之大，可能需要一个NLP专业团队来进行开发，调优。
+2. 应用已有的深度学习模型，直接应用。这个选择比较符合客观需要。我们找到了两个工具包：
+
+    + Tensor2Tensor 工具包：Tensor2Tensor（以下简称T2T）是由 Google Brain 团队使用和维护的开源深度学习模型库，支持多种数据集和模型。T2T 在 github 上有完整的介绍和用法，可以访问这里了解详细信息。
+
+    + Fairseq 工具包：Fairseq 是 Facebook 推出的一个序列建模工具包，这个工具包允许研究和开发人员自定义训练翻译、摘要、语言模型等文本生成任务。这里是它的 PyTorch 实现。
+
+    本案例中，我们使用 T2T 工具包进行模型训练。
+
+
+### 数据收集
+
+有了模型，还需要数据。巧妇难为无米之炊，没有数据，什么都是浮云。数据从哪里来呢？GitHub 上有很多开源贡献者收集和整理了对联数据，可以进行下载使用。
+
+本案例从下面几个渠道获取对联数据：
+1. Github网站上的开源对联数据： https://github.com/wb14123/couplet-dataset/releases
+2. Github网站上的开源古诗数据： https://github.com/chinese-poetry/chinese-poetry
+3. 微软亚洲研究院提供的10万条对联数据（非公开数据）。
+
+### 数据预处理
+
+#### 生成源数据文件
+
+网上提供的对联数据形式各异，需要整理成我们需要的格式。我们创建两个文本文件，命名为 train.txt.up 和 train.txt.down，存放上联和下联数据。每个上联/下联为一行，用换行符 ‘\n’ 分隔。
+
+#### 生成词表文件
+
+接下来我们要统计上下联中出现多少不同的字，用于后续的模型推理。
+1. 将上下联数据每个字以“空格”分隔，合并成一个文件。
+
+    a. 分隔数据的python代码 (split_data.py)：
+    ```
+    import sys
+    filename = sys.argv[1]
+    with open(filename, 'r', encoding='utf-8') as infile:
+    with open(filename + '.clean', 'w', encoding='utf-8') as outfile:
+        lines = infile.readlines()
+        for line in lines:
+        out = ""
+        for i in line.strip():
+            out += i + (' ')
+        out = out[:-1]
+        out += '\n'
+        outfile.write(out)
+    ```
+    b. 执行如下命令完成文件分隔
+    ```
+    python split_data.py train.txt.up
+    python split_data.py train.txt.down
+    ```
+    分隔后生成两个文件：train.txt.up.clean 和 train.txt.down.clean
+
+    c. 合并文件为 merge.txt
+    ```
+    cat train.txt.up train.txt.down > merge.txt
```

Comment:
>train.txt.up train.txt.down [](start = 8, length = 27)

Should we use the .clean files?

Reply:
both are fine for this operation.

Reply:
If we don't use the .clean file, why we need to run step b?

---
In reply to: [279605297](https://github.com/Microsoft/ai-edu/pull/241#discussion_r279605297) [](ancestors = 279605297)